examples/peft_llm_gnn/llm_gnn_model.py

```diff
@@ -115,32 +116,64 @@ def __init__(self, g, node_lm_configs, h_dim, out_dim, num_layers,
                 self._lm_node_feats[ntype] = feats
         self.out = nn.Linear(self.config.hidden_size, out_dim)
         self._loss_fn = gsmodel.ClassifyLossFunc(multilabel=False)
-        #TODO (@qzhuamzn): add initialization for gnn encoding
+        self.gnn = nn.ModuleList()
+        for _ in range(num_layers):
+            self.gnn.append(HeteroGraphConv({
+                _etype: GraphConv(h_dim, h_dim) for _etype in g.etypes
+            }))
+        self.projection = nn.Linear(h_dim, self.config.hidden_size)
+
+    def encode_graph(self, blocks, h):
+        for layer, block in zip(self.gnn, blocks):
+            h = layer(block, h)
+            h = {k: F.relu(v) for k, v in h.items()}
+        src_type, dst_type = blocks[0].ntypes
+        graph_tokens = self.projection(h[dst_type])
+        return graph_tokens
 
     def forward(self, blocks, node_feats, edge_feats, labels, input_nodes):
-        # TODO (qzhuamzn): use GNNs to generate graph tokens
-        h = {}
         output_nodes = blocks[-1].dstdata[dgl.NID]
-
         input_ids = self._lm_node_feats[self.target_ntype][TOKEN_IDX][output_nodes].to(self.llm.device)
         attention_mask = self._lm_node_feats[self.target_ntype][ATT_MASK_IDX][output_nodes].to(self.llm.device)
-        # TODO (qzhuamzn): modify input_ids into input_embeds=[graph_tokens, input_embeds] to support GPEFT
-        model_output = self.llm(input_ids=input_ids, attention_mask=attention_mask)
+
+        graph_tokens = self.encode_graph(blocks, node_feats)
+
+        input_shape = input_ids.size()
+        input_ids = input_ids.view(-1, input_shape[-1])
+        word_embeddings = self.llm.get_input_embeddings()
+        inputs_embeds = torch.cat([graph_tokens.unsqueeze(1), word_embeddings(input_ids)], dim=1)
```

Comment:
Can you add some comments to describe the shape of inputs_embeds and attention_masks and how graph-embeddings are attached into test-embeddings?